### DISCUSSION TITLE
405 Method Not Allowed error on POST requests to Vercel staging environment

### DISCUSSION THREAD

**Discussion by ilcande** [vercel/community#4839] [Help]
### Summary

I am trying to make POST requests to my Vercel staging environment, but I am getting a 405 Method Not Allowed error. All POST requests return 405 errors. Locally the requests work fine, on any Vercel deployment they are failing with 405s. 
I am using a React frontend and a Node.js/Express backend. I have added and enabled CORS, and I have verified that the staging environment is using the correct Node.js version and has the necessary dependencies installed. 

For example, when I try to login an existing user, I am getting the 405 Method Not Allowed error 👇 

![Screenshot 2023-11-12 at 11 24 14](https://github.com/vercel/community/assets/29979860/ebb59b34-7574-4dd9-9687-761ebeba64ce)

I have also tried to addressed this issue following [this thread](https://github.com/orgs/vercel/discussions/2632#discussioncomment-6117635) , but either way I use `axios` or `fetch` I'll get 405s

### Steps to Reproduce

These are code snippets from my server/index.js entrypoint

```javascript
const express = require('express');
const cors = require('cors');
// ...
const app = express();
app.use(cors());
// ...
const login = require('./controllers/login');
// ...
app.post('/signin', validate(loginValidation, {}, {}), login);
```

where the `login` controller  is 
```javascript
const login = async (req, res) => {
  const { email, password } = req.body;

  // ...

  if (!user) {
    return res.status(400).send({ message: 'Email does not exist' });
  } else {
    // .... 
    if (!validPassword) {
      return res.status(400).send({ message: 'Password is not correct' });
    } else {
      // ....

      // Set CORS headers
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE');
      res.setHeader('Access-Control-Allow-Headers', 'Authorization, Content-Type');

      res.send({
        message: 'Login successful',
        // ...
      });
    }
  }
}

module.exports = login;
```

I have tried to set/unset the `headers` in the controller and set/unset the `loginValidation` for staging purposes, but I am not sure where to look next.

Does anyone know what could be causing this error?

**Comment by tomnortonjr** [comment]
I have the same issue in my preview environment. I cannot submit post requests to environments other than production. Curious if this is by design. Works locally, just not deployed to my preview branch.

**Comment by amyegan** [comment]
Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to help with debugging?

**Comment by tomnortonjr** [comment]
I think I identified it as a result of not having a pro plan. Lower
environments are protected and not public.

This makes sense to me.

Thanks,

Tom

**Reply by Sanape** [reply]
so to be able to make post request in vercel i should have a pro plan ?
